torch/distributed/pipelining/schedules.py

```diff
@@ -2097,71 +2166,331 @@ def backward_stage_index(step):
+            ops_count = []
+
+            for i in range(pipeline_parallel_size):
+                temp = {}
+                temp[F] = [0] * 2
+                temp[I] = [0] * 2
+                temp[W] = [0] * 2
+                ops_count.append(temp)
+
+            fbw_mem = [3, -1, -2]
+            max_mem = 3 * (pipeline_parallel_size * 2)
+            end_time = [-1] * n_node
+            cur_time = [0] * pipeline_parallel_size
+            mem = [0] * pipeline_parallel_size
+            rank_bubble = [0] * pipeline_parallel_size
+            pending_w: List[Deque[Tuple[int, int, int]]] = [
+                deque() for _ in range(pipeline_parallel_size)
+            ]
+            schedule: Dict[int, List[Tuple[int, _ComputationType, int]]] = {}
+            for i in range(pipeline_parallel_size):
+                schedule[i] = []
+
+            def put_w(rank):
+                assert len(pending_w[rank]) > 0
+                _, chunk_, _ = pending_w[rank].popleft()
+                put(W, chunk_, rank)
+
+            def put(op, chunk, rank, assert_cnt=True):
+                """
+                Schedules an operation (forward, backward, or weight update) for a given chunk and rank.
+                op (_ComputationType): The type of operation (FORWARD, BACKWARD, WEIGHT).
+                chunk (int): The model chunk index (0 or 1).
+                rank (int): The pipeline stage index.
+                """
+                _cur_time = _no_bubble = cur_time[rank] + 1
+                category = category_map[op]
+                # Number of ops (F/B/W) with the same (action, chunk) on current rank
+                _op_count = ops_count[rank][op][chunk]
+                if chunk == chunk_1 and op in (F, I):
+                    if rank < pipeline_parallel_size - 1:
+                        _fa_id = get_id(op, chunk, rank + 1, microbatch_id=_op_count)
+                        assert end_time[_fa_id] >= 0
+                        _cur_time = max(_cur_time, end_time[_fa_id] + 1)
+                if chunk == chunk_0 and op in (F, I):
+                    if rank > 0:
+                        _fa_id = get_id(op, chunk, rank - 1, microbatch_id=_op_count)
+                        assert end_time[_fa_id] >= 0, f"{category}, {chunk}, {rank}, {_op_count}"
+                        _cur_time = max(_cur_time, end_time[_fa_id] + 1)
+                _id = get_id(op, chunk, rank, microbatch_id=_op_count)
+                if ops_count[rank][F][chunk_0] > 0:
+                    rank_bubble[rank] += _cur_time - _no_bubble
+                end_time[_id] = _cur_time
+                cur_time[rank] = _cur_time
+                mem[rank] += fbw_mem[category]
+                # For BACKWARD and WEIGHT operation, we will schedule chunk 1 before 0 so
+                # inversing the order before adding to the schedule
+                temp_chunk = chunk if op == F else 1 - chunk
+                schedule[rank].append((_op_count, op, temp_chunk))
```

Comment:
nit: could you directly construct an _Action() object here as you append, instead of creating a similar but not the same representation?  Otoh I understand that for the logic here it is convenient to refer to local chunk ID (0,1) rather than global stage_id which combines local chunk_id with rank.  IIUC op_count is the same as microbatch_id?